external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala

```diff
@@ -202,34 +203,40 @@ private[sql] object AvroUtils extends Logging {
     }
   }
 
-  /**
-   * Extract a single field from `avroSchema` which has the desired field name,
-   * performing the matching with proper case sensitivity according to [[SQLConf.resolver]].
-   *
-   * @param avroSchema The schema in which to search for the field. Must be of type RECORD.
-   * @param name The name of the field to search for.
-   * @param avroPath The seq of parent field names leading to `avroSchema`.
-   * @return `Some(match)` if a matching Avro field is found, otherwise `None`.
-   * @throws IncompatibleSchemaException if `avroSchema` is not a RECORD or contains multiple
-   *                                     fields matching `name` (i.e., case-insensitive matching
-   *                                     is used and `avroSchema` has two or more fields that have
-   *                                     the same name with difference case).
-   */
-  private[avro] def getAvroFieldByName(
-      avroSchema: Schema,
-      name: String,
-      avroPath: Seq[String]): Option[Schema.Field] = {
+  class AvroSchemaHelper(avroSchema: Schema, avroPath: Seq[String]) {
     if (avroSchema.getType != Schema.Type.RECORD) {
       throw new IncompatibleSchemaException(
         s"Attempting to treat ${avroSchema.getName} as a RECORD, but it was: ${avroSchema.getType}")
     }
-    avroSchema.getFields.asScala.filter(f => SQLConf.get.resolver(f.name(), name)).toSeq match {
-      case Seq(avroField) => Some(avroField)
-      case Seq() => None
-      case matches => throw new IncompatibleSchemaException(s"Searching for '$name' in Avro " +
+
+    val schemaMap = avroSchema.getFields.asScala.groupBy { f =>
+        f.name.toLowerCase(Locale.ROOT)
+      }.map { case (k, v) =>
+      (k, v.toSeq) // needed for scala 2.13
+    }
+
+    /**
+     * Extract a single field from the contained avro schema which has the desired field name,
+     * performing the matching with proper case sensitivity according to SQLConf.resolver.
+     *
+     * @param name The name of the field to search for.
+     * @return `Some(match)` if a matching Avro field is found, otherwise `None`.
+     */
+    def getFieldByName(name: String): Option[Schema.Field] = {
+
+      // get candidates, ignoring case of field name
+      val candidates = schemaMap.get(name.toLowerCase(Locale.ROOT))
+        .getOrElse(Seq.empty[Schema.Field])
+
+      // search candidates, taking into account case sensitivity settings
+      candidates.filter(f => SQLConf.get.resolver(f.name(), name)) match {
+        case Seq(avroField) => Some(avroField)
```

Comment:
Can you provide any commentary on why you chose to always lowercase and then resolve later based on case sensitivity, vs. storing the map keys as case insensitive or sensitive based on the SQL conf? Is it so that we can make use of `SQLConf#resolver` to perform the resolution instead of checking the value of the case sensitive conf ourselves?

Reply:
>Is it so that we can make use of SQLConf#resolver to perform the resolution

Well, that was certainly one reason (so I don't have to decide the resolver, in case that decision logic changes).

But I also wanted to keep the code similar to the original implementation (i.e., duplicates are detected at lookup time, etc.).

I suppose my version wouldn't help much in the case of a wide table where all fields map to the same lower case name (e.g., a 10 character field name that just repeats with many different permutations of case). Hopefully that is an extremely unusual situation.

Reply:
@xkrogen @gengliangwang What do you think (referring to xkrogen's comment above)? Should we instead build a map of `String -> Schema.Field` rather than `String -> Seq[Schema.Field]`, using the value of spark.sql.caseSensitive to determine whether the keys are mixed case or single case (and checking for duplicates, in the case insensitive mode, at Map build time, rather than lookup time)?

Reply:
I feel the current approach is a bit simpler (more of the code is shared between case-sensitive vs. insensitive), with only very minor performance difference. So even though it was my suggestion to change it, I'm leaning towards the current implementation :)

Reply:
The current approach LGTM.